File: 1. Basic Topics/2. Backtracking/py/39. Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        tmp = []
        candidates.sort() 
        self.backtrack(candidates, target, ans, tmp, 0)
        return ans

    def backtrack(self, candidates, target, ans, tmp, start):
        if target == 0:
            ans.append(tmp.copy())  
            return
        for i in range(start, len(candidates)):
            if target < candidates[i]:
                break
            tmp.append(candidates[i])
            self.backtrack(candidates, target - candidates[i], ans, tmp, i)
            tmp.pop()

File: 1. Basic Topics/2. Backtracking/cpp/39. Combination Sum.cpp
class Solution {
public:
    void backtrack(vector<int>& candidates, int target, vector<vector<int>> &ans, vector<int> &tmp, int start){
        if(target == 0){
            ans.push_back(tmp);
            return;
        }
        for(int i = start; i<candidates.size() && target >=candidates[i]; i++){
            tmp.push_back(candidates[i]);
            backtrack(candidates, target-candidates[i], ans, tmp, i);
            tmp.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> tmp;
        backtrack(candidates, target, ans, tmp, 0);
        return ans;
    }
};